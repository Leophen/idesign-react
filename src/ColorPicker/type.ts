export interface colorListType {
  value: string
}

export interface ColorPickerProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 触发颜色块类名
   */
  triggerClassName?: string;
  /**
   * 触发颜色块样式
   */
  triggerStyle?: React.CSSProperties;
  /**
   * 触发颜色块尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 固定颜色值（受控）
   */
  value?: string;
  /**
   * 默认颜色值（非受控）
   * @default #5e62ea
   */
  defaultValue?: string;
  /**
   * 默认可选值
   */
  colorList?: colorListType[]
  /**
   * 修改颜色值时触发
   */
  onChange?: (val: string) => void
  /**
   * 切换颜色面板时触发
   */
  onTrigger?: (val: string, visible: boolean) => void
}

export interface ColorPanelProps extends ColorPickerProps {
  /**
   * 点击关闭按钮时触发
   */
  onClose?: () => void
}

export interface ColorItemProps {
  /**
   * 颜色
   * @default #5e62ea
   */
  color?: string;
  /**
   * 点击事件
   */
  onClick?: (val: string) => void;
}

export interface ColorCursorProps {
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 初始横坐标比例
   */
  x?: number;
  /**
   * 初始纵坐标比例
   */
  y?: number;
  /**
   * 拖拽方向
   * @default xy
   */
  mode?: 'x' | 'y' | 'xy';
  /**
   * 滑块颜色
   * @default #5e62ea
   */
  color?: string;
}
