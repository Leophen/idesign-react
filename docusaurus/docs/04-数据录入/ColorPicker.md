---
title: ColorPicker 颜色选择器
---

颜色选择器，用于颜色选择的操作控件,相比较下拉选择框而言,颜色选择器不仅可以对颜色进行选择,也可以对颜色进一步编辑。 颜色选择器，用于颜色选择和调整的选择器控件。

## 引入

```js
import {ColorPicker} from 'idesign';
```

## 基本用法

```jsx live
function App() {
    const [color, setColor] = useState('#4D4DFF');

    const handleChange = color => {
        console.log(color);
        setColor(color);
    };

    return (
        <div className="demo-column">
            <h5>无默认值</h5>
            <ColorPicker />

            <h5>有默认值（非受控）</h5>
            <ColorPicker defaultColor={color} />

            <h5>有固定值（受控?）</h5>
            <ColorPicker color={color} />

            <h5>通用方法</h5>
            <ColorPicker color={color} onChange={handleChange} />
        </div>
    );
}
```

## 禁用状态

```jsx live
<ColorPicker defaultColor='#4D4DFF' disabled />
```

## 可清空颜色值

```jsx live
<ColorPicker defaultColor='#4D4DFF' allowClear />
```

## 简洁颜色选择器

简洁颜色选择器简化了选择器的部分，只显示已选颜色。

```jsx live
<ColorPicker defaultColor='#4D4DFF' type="simple" />
```

## 混合颜色选择器

相对基本颜色选择器而言,混合颜色选择器针对更加专业的颜色调节的业务场景，例如图文编辑中图片颜色、文字颜色调节等场景。

```jsx live
<ColorPicker defaultColor='#4D4DFF' type="custom" />
```

## API

| 参数             | 说明                                                                                 | 类型                           | 默认值                         |
| ---------------- | ------------------------------------------------------------------------------------ | ------------------------------ | ------------------------------ |
| defaultColor     | 默认选中颜色                                                                         | `string`                       | `--`                           |
| defaultColors    | 默认色板，常规/推荐颜色，可以快速选择颜色，如果需要无颜色选项，设置为transparent即可 | `string[]`                     | `DEFAULT_COLORS(具体值见下方)` |
| disabled         | 是否可选                                                                             | `boolean`                      | `false`                        |
| type             | 颜色选择器有三种类型                                                                 | `simple`〡`cusAtom`〡`default` | `default`                      |
| color            | 指定当前选中的色值，HEX格式的颜色                                                    | `string`                       | `--`                           |
| placeholder      | 当前下拉选择器的placeholder                                                          | `string`                       | `--`                           |
| inputPlaceholder | 当type设置为custom时，下拉中输入框的placeholder                                      | `string`                       | `--`                           |
| allowClear       | 支持清除                                                                             | `boolean`                      | `false`                        |
| onChange         | 选中色值即色值变化时，调用该函数。其中color: {hex, rgb, hsl, hsv}                    | `function(hex〡color)`         | `--`                           |

```js
const DEFAULT_COLORS = [
    'transparent',
    '#FFFFFF',
    '#FFC27D',
    '#FF908C',
    '#7BD964',
    '#57D9D4',
    '#7DADFF',
    '#9497FF',
    '#E62E6B',
    '#FAD000',
    '#FF9326',
    '#F33E3E',
    '#30BF13',
    '#30B5F2',
    '#2468F2',
    '#4D4DFF'
];
```
