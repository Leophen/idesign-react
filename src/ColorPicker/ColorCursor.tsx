import React, { useState, useRef, useEffect } from 'react';
import './index.scss';
import { ColorCursorProps } from './type';

const ColorCursor: React.FC<ColorCursorProps> = (props) => {
  const {
    x = 0,
    y = 0,
    mode = 'xy',
    color,
    style
  } = props

  const cursor = useRef<any>(null)

  const [translate, setTranslate] = useState({
    x: 0,
    y: 0
  })

  const [parent, setParent] = useState({
    width: 0,
    height: 0
  })

  useEffect(() => {
    parent.width = cursor.current.parentNode.clientWidth
    parent.height = cursor.current.parentNode.clientHeight
    setParent({ ...parent })
  }, [])

  useEffect(() => {
    if (mode === 'x') {
      translate.x = x * (parent.width - 12)
    } else if (mode === 'y') {
      translate.y = y * (parent.height - 12)
    } else {
      translate.x = (x * parent.width) - 6
      translate.y = (y * parent.height) - 6
    }
    setTranslate({ ...translate })
  }, [x, y])

  return (
    <div
      className="i-color-picker__cursor"
      ref={cursor}
      style={{
        ...(style || {}),
        ...{
          transform: `translate(${translate.x}px, ${translate.y}px)`,
          background: color
        }
      }}
    />
  )
}

export default ColorCursor;
