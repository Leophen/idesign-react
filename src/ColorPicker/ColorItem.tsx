import React from 'react';
import './index.scss';
import { ColorItemProps } from './type';

const ColorItem: React.FC<ColorItemProps> = (props) => {
  const {
    color = '#265CF0',
    onClick
  } = props

  const handleClick = () => {
    onClick?.(color)
  }

  return (
    <div
      className="i-color-panel-color"
      style={{
        background: color
      }}
      onClick={handleClick}
    />
  )
}

export default ColorItem;
