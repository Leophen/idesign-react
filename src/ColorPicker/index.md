---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 9
---

# ColorPicker 颜色选择器

用于颜色选择的组件，支持 `Hex` 和 `RGB` 格式。

## 基本用法

可通过 `value` 设置实时颜色值，通过 `onChange` 设置修改颜色值触发的方法，通过 `onTrigger` 设置切换颜色面板显示隐藏时触发的方法。

```tsx
import React, { useState } from 'react'
import { ColorPicker } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('#4362E2')

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  const handleTrigger = (val, visible) => {
    console.log(val, visible)
  }

  return (
    <>
      <h4>无默认值</h4>
      <ColorPicker />
      <h4>有默认值</h4>
      <ColorPicker defaultValue={value} />
      <h4>有受控值</h4>
      <ColorPicker value={value} />
      <h4>一般用法</h4>
      <ColorPicker
        value={value}
        onChange={handleChange}
        onTrigger={handleTrigger}
      />
    </>
  )
}

export default App
```

## 不同尺寸

提供 `small`、`medium`（默认）、`large` 三种尺寸的颜色触发块。

```tsx
import React from 'react';
import { ColorPicker } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <ColorPicker size="small" />
        <ColorPicker size="medium" />
        <ColorPicker size="large" />
      </div>
    </div>
  );
};

export default App;
```

## 预设颜色

可通过 `colorList` 属性传入颜色数组设置预设颜色。

```tsx
import React from 'react';
import { ColorPicker } from 'idesign-react';

const defaultColor = [
  {
    value: 'rgb(206, 55, 46)'
  },
  {
    value: 'rgb(245, 192, 66)'
  },
  {
    value: 'rgb(92, 192, 131)'
  },
  {
    value: 'rgb(86, 116, 245)'
  },
  {
    value: 'rgb(103, 98, 192)'
  },
]

const App = () => {
  return (
    <ColorPicker colorList={defaultColor} />
  );
};

export default App;
```

## 颜色面板

可通过 `<ColorPicker.Panel />` 单独使用颜色面板。

```tsx
import React from 'react';
import { ColorPicker } from 'idesign-react';

const App = () => {
  return (
    <ColorPicker.Panel />
  );
};

export default App;
```

<API />
