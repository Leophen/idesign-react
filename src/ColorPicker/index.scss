.i-color {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
  border: 1px solid var(--i-border);
  border-radius: 6px;
  cursor: pointer;
  &.i-color--size-small {
    width: 24px;
    height: 24px;
  }
  &.i-color--size-large {
    width: 40px;
    height: 40px;
  }
}

.i-color-panel {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 240px;
}

.i-color-panel-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  padding: 0 16px;
  color: var(--i-font-2);
}

.i-color-panel-header-txt {
  cursor: default;
}

.i-color-panel-header-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  margin-right: -8px;
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-hover);
  }
}

.i-color-panel-block {
  position: relative;
  width: 240px;
  height: 240px;
}

.i-color-panel-block__white,
.i-color-panel-block__black {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.i-color-panel-block__white {
  background: linear-gradient(90deg, #fff, rgba(255, 255, 255, 0));
}

.i-color-panel-block__black {
  background: linear-gradient(0deg, #000, transparent);
}

.i-color-picker__cursor {
  position: absolute;
  z-index: 3;
  box-sizing: border-box;
  width: 12px;
  height: 12px;
  border: 2px solid #fff;
  border-radius: 6px;
  box-shadow: rgba(0, 0, 0, 0.2) 0px 0px 0px 0.6px;
  user-select: none;
  pointer-events: none;
  -webkit-user-drag: none;
}

.i-color-panel-controls {
  display: flex;
  gap: 8px;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  padding: 16px;
}

.i-color-panel-controls__dropper,
.i-color-panel-controls__bar {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.i-color-panel-controls__dropper {
  width: 32px;
  height: 32px;
  border: 1px solid var(--i-border);
  border-radius: 16px;
  cursor: pointer;
  transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1);
  .icon-Dropper {
    color: var(--i-font-4);
  }
  &:hover {
    background: var(--i-bg-hover);
  }
}

.i-color-panel-controls__bar {
  flex: 1;
}

.i-color-panel-bar__rgb,
.i-color-panel-bar__a {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 12px;
  border-radius: 6px;
  box-shadow: rgb(0 0 0 / 20%) 0px 0px 0px 0.6px inset;
}

.i-color-panel-bar__rgb {
  margin-bottom: 8px;
  background: linear-gradient(
    90deg,
    #f00 0,
    #ff0 17%,
    #0f0 33%,
    #0ff 50%,
    #00f 67%,
    #f0f 83%,
    #f00
  );
}

.i-color-panel-bar__a {
  position: relative;
  overflow: hidden;
}

.i-color-panel-bar__a-color,
.i-color-panel-bar__a-bg {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.i-color-panel-bar__a-color {
  z-index: 2;
  border-radius: 6px;
  box-shadow: rgb(0 0 0 / 20%) 0px 0px 0px 0.6px inset;
}

.i-color-panel-bar__a-bg {
  z-index: 1;
  background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAIAAADZF8uwAAAAGUlEQVQYV2M4gwH+YwCGIasIUwhT25BVBADtzYNYrHvv4gAAAABJRU5ErkJggg==');
}

.i-color-panel-values {
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  padding: 0px 16px 16px;
}

.i-color-panel-input__wrapper {
  display: flex;
  align-items: center;
  box-sizing: border-box;
  width: 66%;
  overflow: hidden;
  border: 1px solid var(--i-border);
  border-radius: 4px;
  transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1);
  &:focus-within {
    border-color: var(--i-primary);
  }
  .i-input {
    border: none;
    border-right: 1px solid var(--i-border);
    border-radius: 0;
    &:hover,
    &:focus-within {
      background-color: var(--i-bg-hover);
    }
  }
  .i-input__inner {
    padding-right: 0;
    &::selection {
      background: transparent;
    }
    &:focus {
      &::selection {
        background: var(--i-primary-bg-2);
      }
    }
  }
}

.i-color-panel-input__class {
  display: flex;
  flex: 75%;
  align-items: center;
}

.i-color-panel-input__alpha {
  flex: 25%;
  .i-input {
    border-right: none;
  }
}

.i-color-panel-footer {
  display: flex;
  flex-wrap: wrap;
  gap: 8px;
  box-sizing: border-box;
  width: 100%;
  padding: 16px;
  border-top: 1px solid var(--i-border);
}

.i-color-panel-color {
  width: 16px;
  height: 16px;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: rgb(0 0 0 / 20%) 0px 0px 0px 0.6px inset;
  cursor: pointer;
}
