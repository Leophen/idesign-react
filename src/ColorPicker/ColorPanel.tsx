import React, { useState, useRef, useEffect } from 'react';
import './index.scss';
import tinycolor from 'tinycolor2'
import Select from '../Select';
import Input from '../Input';
import Icon from '../Icon';
import ColorCursor from './ColorCursor';
import ColorItem from './ColorItem';
import { defaultColor } from './index'
import { ColorPanelProps } from './type';

const ColorPanel: React.FC<ColorPanelProps> = (props) => {
  const {
    value = '#265CF0',
    colorList = defaultColor,
    onChange,
    onClose
  } = props;

  // 颜色值
  const [colors, setColors] = useState({
    rgbVal: tinycolor(value).toRgbString(),
    hexVal: tinycolor(value).toHexString(),
    r: tinycolor(value).toRgb().r,
    g: tinycolor(value).toRgb().g,
    b: tinycolor(value).toRgb().b,
    h: tinycolor(value).toHsv().h,
    s: tinycolor(value).toHsv().s,
    v: tinycolor(value).toHsv().v,
    a: tinycolor(value).getAlpha()
  })

  // 是否为移动状态
  const handleMoving = (ifMoving: boolean) => {
    if (ifMoving) {
      document.body.style.userSelect = 'none'
    } else {
      document.body.style.userSelect = ''
    }
  }

  // 调色板、色阶柱、透明度柱 节点宽高位置
  const [rect, setRect] = useState({
    panel: {
      width: 0,
      height: 0,
      left: 0,
      top: 0
    },
    rgb: {
      width: 0,
      left: 0
    },
    a: {
      width: 0,
      left: 0
    },
  })
  const panelNode = useRef<HTMLDivElement>(null)
  const rgbBarNode = useRef<HTMLDivElement>(null)
  const aBarNode = useRef<HTMLDivElement>(null)
  useEffect(() => {
    // 初始化给各节点宽高及位置参数赋值（popup 打开后再执行）
    setTimeout(() => {
      const panelRect = panelNode.current?.getBoundingClientRect()
      rect.panel.width = panelRect?.width || 0
      rect.panel.height = panelRect?.height || 0
      rect.panel.left = panelRect?.left || 0
      rect.panel.top = panelRect?.top || 0
      const rgbRect = rgbBarNode.current?.getBoundingClientRect()
      rect.rgb.width = rgbRect?.width || 0
      rect.rgb.left = rgbRect?.left || 0
      const aRect = aBarNode.current?.getBoundingClientRect()
      rect.a.width = aRect?.width || 0
      rect.a.left = aRect?.left || 0
      setRect({ ...rect })
    })
    // 设置初始滑块位置
    const currentColor = tinycolor(colors.rgbVal)
    location.panel.x = currentColor.toHsv().s
    location.panel.y = 1 - currentColor.toHsv().v
    location.rgb.x = currentColor.toHsv().h / 360
    location.a.x = currentColor.getAlpha()
    setLocation({ ...location })
  }, [])

  // 调色板、色阶柱、透明度柱 坐标值
  const [location, setLocation] = useState({
    panel: {
      x: 0,
      y: 0
    },
    rgb: {
      x: 0
    },
    a: {
      x: 0
    }
  })

  const emitColor = (tinyObj: any) => {
    if (colorType === 'hex') {
      onChange?.(tinyObj.toHex8String())
    } else {
      onChange?.(tinyObj.toRgbString())
    }
  }

  // 通用更新滑块位置函数
  const updateCursorLocation = (tinyObj: any) => {
    location.panel.x = tinyObj.toHsv().s
    location.panel.y = 1 - tinyObj.toHsv().v
    location.rgb.x = tinyObj.toHsv().h / 360
    location.a.x = tinyObj.getAlpha()
    setLocation({ ...location })
  }

  // 传入一种颜色值 -> 更新全部颜色值
  const updateColor = (color: string, alpha: number) => {
    const tinyObj = tinycolor(color)
    tinyObj.setAlpha(alpha)

    colors.rgbVal = tinyObj.toRgbString()
    colors.hexVal = tinyObj.toHexString()
    colors.r = tinyObj.toRgb().r
    colors.g = tinyObj.toRgb().g
    colors.b = tinyObj.toRgb().b
    colors.s = tinyObj.toHsv().s
    colors.v = tinyObj.toHsv().v
    colors.a = alpha

    setColors({ ...colors })
    emitColor(tinyObj)
  }

  // 传入调色板坐标 -> 更新颜色
  const updatePanelColor = (x: number, y: number) => {
    const hsv = `hsv(${colors.h.toFixed(0)}, ${(x * 100).toFixed(0)}%, ${((1 - y) * 100).toFixed(0)}%)`
    updateColor(hsv, colors.a)
    // 更新滑块位置
    location.panel.x = x
    location.panel.y = y
    setLocation({ ...location })
  }

  // 传入色阶柱坐标 -> 更新颜色
  const updateRgbColor = (x: number) => {
    let currentX = x
    currentX === 1 && (currentX = 0)  // 左右极限值去重
    const currentHue = Math.round((currentX) * 360 * 100) / 100
    // 单独更新色阶
    colors.h = currentHue
    setColors({ ...colors })
    // 更新全部颜色
    const hsv = `hsv(${currentHue}, ${colors.s}, ${colors.v})`
    updateColor(hsv, colors.a)
    // 更新滑块位置
    location.rgb.x = x
    setLocation({ ...location })
  }

  // 传入透明度柱坐标 -> 更新颜色
  const updateAColor = (x: number) => {
    let currentX = Number(x.toFixed(2))
    updateColor(colors.rgbVal, currentX)
    // 更新滑块位置
    location.a.x = x
    setLocation({ ...location })
  }

  // 移动调色板
  const handlePanelMove = (e: any) => {
    let moveX = e.clientX - rect.panel.left
    let moveY = e.clientY - rect.panel.top
    const maxX = rect.panel.width
    const maxY = rect.panel.height
    const minX = 0
    const minY = 0
    moveX < (minX) && (moveX = (minX))
    moveX > (maxX) && (moveX = (maxX))
    moveY < (minY) && (moveY = (minY))
    moveY > (maxY) && (moveY = (maxY))
    updatePanelColor(moveX / maxX, moveY / maxY)
  }
  const handlePanelUp = () => {
    handleMoving(false)
    window.removeEventListener('mousemove', handlePanelMove);
    window.removeEventListener('mouseup', handlePanelUp);
  };

  // 移动色阶柱
  const handleRgbMove = (e: any) => {
    let moveX = e.clientX - rect.rgb.left
    const maxX = rect.rgb.width
    const minX = 0
    moveX < (minX) && (moveX = (minX))
    moveX > (maxX) && (moveX = (maxX))
    updateRgbColor(moveX / maxX)
  }
  const handleRgbUp = () => {
    handleMoving(false)
    window.removeEventListener('mousemove', handleRgbMove);
    window.removeEventListener('mouseup', handleRgbUp);
  };

  // 移动透明度柱
  const handleAMove = (e: any) => {
    let moveX = e.clientX - rect.a.left
    const maxX = rect.a.width
    const minX = 0
    moveX < (minX) && (moveX = (minX))
    moveX > (maxX) && (moveX = (maxX))
    updateAColor(moveX / maxX)
  }
  const handleAUp = () => {
    handleMoving(false)
    window.removeEventListener('mousemove', handleAMove);
    window.removeEventListener('mouseup', handleAUp);
  };

  const handleUsualDown = (e: React.MouseEvent, type: 'panel' | 'rgb' | 'a') => {
    e.persist();
    let downX = 0
    let downY = 0
    handleMoving(true)
    if (type === 'panel') {
      // 点击调色板 -> 更新颜色
      downX = e.clientX - rect.panel.left
      downY = e.clientY - rect.panel.top
      updatePanelColor(downX / rect.panel.width, downY / rect.panel.height)
      // 移动调色板 -> 更新颜色
      window.addEventListener('mousemove', handlePanelMove);
      window.addEventListener('mouseup', handlePanelUp);
    } else if (type === 'rgb') {
      // 点击色阶柱 -> 更新颜色
      downX = e.clientX - rect.rgb.left
      updateRgbColor(downX / rect.rgb.width)
      // 移动色阶柱 -> 更新颜色
      window.addEventListener('mousemove', handleRgbMove);
      window.addEventListener('mouseup', handleRgbUp);
    } else {
      // 点击透明度柱 -> 更新颜色
      downX = e.clientX - rect.a.left
      updateAColor(downX / rect.a.width)
      // 移动透明度柱 -> 更新颜色
      window.addEventListener('mousemove', handleAMove);
      window.addEventListener('mouseup', handleAUp);
    }
  }

  // 通过更新颜色函数
  const handleUsualUpdate = (val: string) => {
    const currentColor = tinycolor(val)
    updateColor(val, currentColor.getAlpha())
    // 单独更新色阶
    colors.h = currentColor.toHsv().h
    setColors({ ...colors })
    // 更新滑块位置
    updateCursorLocation(currentColor)
  }

  // 选择颜色值类型
  const [colorType, setColorType] = useState('hex')
  const handleSelect = (val: any) => {
    setColorType(val)
  }

  // RGBA 输入框变化时触发
  const inputChange = (val: string, type: 'r' | 'g' | 'b') => {
    let currentVal = Number(val)
    let color = ''
    if (type === 'r') {
      colors.r = currentVal
      color = `rgba(${val}, ${colors.g}, ${colors.b}, ${colors.a})`
    } else if (type === 'g') {
      colors.g = currentVal
      color = `rgba(${colors.r}, ${val}, ${colors.b}, ${colors.a})`
    } else {
      colors.b = currentVal
      color = `rgba(${colors.r}, ${colors.g}, ${val}, ${colors.a})`
    }
    setColors({ ...colors })
    handleUsualUpdate(color)
  }
  const inputChangeR = (val: string) => {
    inputChange(val, 'r')
  }
  const inputChangeG = (val: string) => {
    inputChange(val, 'g')
  }
  const inputChangeB = (val: string) => {
    inputChange(val, 'b')
  }
  const inputChangeA = (val: string) => {
    let currentVal = Number(val) / 100
    colors.a = currentVal
    setColors({ ...colors })
    updateAColor(currentVal)
  }

  const hexBackup = useRef('')
  const inputFocusHex = (val: string) => {
    hexBackup.current = val
  }
  const inputChangeHex = (val: string) => {
    colors.hexVal = val
    setColors({ ...colors })
  }
  const inputBlurHex = (val: string) => {
    const currentColor = tinycolor(val)
    currentColor.setAlpha(colors.a)
    if (currentColor.isValid()) {
      updateColor(val, colors.a)
      // 单独更新色阶
      colors.h = currentColor.toHsv().h
      setColors({ ...colors })
      // 更新滑块位置
      updateCursorLocation(currentColor)
    } else {
      colors.hexVal = hexBackup.current
      setColors({ ...colors })
    }
  }

  // 取色
  const handleClickDropper = async () => {
    // @ts-ignore
    const eyeDropper = new EyeDropper();
    const result = await eyeDropper.open();
    const color = result.sRGBHex

    handleUsualUpdate(color)
  }

  return (
    <div className="i-color-panel">
      <header className="i-color-panel-header">
        <div className="i-color-panel-header-txt">
          颜色选择器
        </div>
        <div
          className="i-color-panel-header-icon"
          onClick={() => {
            onClose?.()
          }}
        >
          <Icon name="Close" />
        </div>
      </header>

      <section
        className="i-color-panel-block"
        style={{ background: `hsl(${colors.h}, 100%, 50%)` }}
      >
        <div className="i-color-panel-block__white" />
        <div
          className="i-color-panel-block__black"
          ref={panelNode}
          onMouseDown={(e) => handleUsualDown(e, 'panel')}
        />
        <ColorCursor
          x={location.panel.x}
          y={location.panel.y}
          color={colors.rgbVal}
        />
      </section>

      <section className="i-color-panel-controls">
        {/* @ts-ignore */}
        {!!window.EyeDropper &&
          <div
            className="i-color-panel-controls__dropper"
            onClick={handleClickDropper}
          >
            <Icon name="Dropper" />
          </div>
        }
        <div className="i-color-panel-controls__bar">
          <div
            className="i-color-panel-bar__rgb"
            ref={rgbBarNode}
            onMouseDown={(e) => handleUsualDown(e, 'rgb')}
          >
            <ColorCursor
              x={location.rgb.x}
              mode="x"
            />
          </div>
          <div
            className="i-color-panel-bar__a"
            ref={aBarNode}
            onMouseDown={(e) => handleUsualDown(e, 'a')}
          >
            <ColorCursor
              mode="x"
              x={location.a.x}
              style={{ background: 'rgba(0, 0, 0, 0.4)' }}
            />
            <section
              className="i-color-panel-bar__a-color"
              style={{ background: `linear-gradient(90deg, rgba(255, 0, 0, 0) 0%, hsl(${colors.h}, 100%, 50%) 100%)` }}
            />
            <section className="i-color-panel-bar__a-bg"></section>
          </div>
        </div>
      </section>

      <section className="i-color-panel-values">
        <Select
          width={60}
          value={colorType}
          size="small"
          clearable={false}
          onChange={handleSelect}
        >
          <Select.Item value="hex">Hex</Select.Item>
          <Select.Item value="rgb">RGB</Select.Item>
        </Select>
        <div className="i-color-panel-input__wrapper">
          <div className="i-color-panel-input__class">
            {colorType === 'hex' ?
              <Input
                value={colors.hexVal}
                size="small"
                onFocus={inputFocusHex}
                onChange={inputChangeHex}
                onBlur={inputBlurHex}
              /> : <>
                <Input
                  value={colors.r.toFixed(0)}
                  type="number"
                  size="small"
                  maxNumber={255}
                  minNumber={0}
                  selectAll
                  hideNumberBtn
                  onChange={inputChangeR}
                />
                <Input
                  value={colors.g.toFixed(0)}
                  type="number"
                  size="small"
                  maxNumber={255}
                  minNumber={0}
                  selectAll
                  hideNumberBtn
                  onChange={inputChangeG}
                />
                <Input
                  value={colors.b.toFixed(0)}
                  type="number"
                  size="small"
                  maxNumber={255}
                  minNumber={0}
                  selectAll
                  hideNumberBtn
                  onChange={inputChangeB}
                />
              </>
            }
          </div>
          <div className="i-color-panel-input__alpha">
            <Input
              value={(colors.a * 100).toFixed(0)}
              type="number"
              size="small"
              maxNumber={100}
              minNumber={0}
              selectAll
              hideNumberBtn
              onChange={inputChangeA}
            />
          </div>
        </div>
      </section>

      <footer className="i-color-panel-footer">
        {colorList.map(item =>
          <ColorItem color={item?.value} key={item?.value} onClick={() => handleUsualUpdate(item?.value)} />
        )}
      </footer>
    </div>
  );
};

export default ColorPanel;
