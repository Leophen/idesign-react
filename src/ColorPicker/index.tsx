import React, { useState, useRef, useMemo } from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import Popup from '../Popup';
import ColorPanel from './ColorPanel';
import { ColorPickerProps } from './type';

export const defaultColor = [
  {
    value: 'rgb(206, 55, 46)'
  },
  {
    value: 'rgb(237, 149, 160)'
  },
  {
    value: 'rgb(252, 238, 239)'
  },
  {
    value: 'rgb(232, 109, 44)'
  },
  {
    value: 'rgb(245, 192, 66)'
  },
  {
    value: 'rgb(238, 209, 103)'
  },
  {
    value: 'rgb(127, 225, 89)'
  },
  {
    value: 'rgb(114, 212, 183)'
  },
  {
    value: 'rgb(135, 214, 230)'
  },
  {
    value: 'rgb(216, 238, 242)'
  },
  {
    value: 'rgb(86, 116, 245)'
  },
  {
    value: 'rgb(103, 98, 192)'
  },
  {
    value: 'rgb(223, 221, 252)'
  },
  {
    value: 'rgb(92, 192, 131)'
  },
  {
    value: 'rgb(210, 90, 182)'
  },
]

const ColorPicker: React.FC<ColorPickerProps> & { Panel: React.ElementType } = (props) => {
  const {
    className,
    style,
    triggerClassName,
    triggerStyle,
    size,
    value,
    defaultValue = '#265CF0',
    colorList = defaultColor,
    onChange,
    onTrigger,
    ...restProps
  } = props

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);
  const [visible, setVisible] = useState(false)

  const currentColor = useRef(innerValue)
  const handleChange = (val: string) => {
    setInnerValue(val)
    currentColor.current = val
  }

  const popupChange = (val: boolean) => {
    if (val) {
      document.documentElement.style.overflowY = 'hidden';
    } else {
      document.documentElement.style.overflowY = '';
    }
    setVisible(val)
    onTrigger?.(currentColor.current, val)
  }

  const handleClose = () => {
    setVisible(false)
    onTrigger?.(currentColor.current, false)
  }

  const colorPanel = useMemo(() =>
    <ColorPanel
      value={innerValue}
      colorList={colorList}
      onChange={handleChange}
      onClose={handleClose}
    />, [visible]
  )

  return (
    <div
      className={classNames(
        'i-color-picker',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      <Popup
        visible={visible}
        content={colorPanel}
        trigger="click"
        placement='bottom-left'
        onTrigger={popupChange}
      >
        <div
          className={classNames(
            'i-color',
            size && `i-color--size-${size}`,
            triggerClassName
          )}
          style={{ ...triggerStyle, background: innerValue }}
        ></div>
      </Popup>
    </div>
  )
}

ColorPicker.Panel = ColorPanel;

ColorPanel.displayName = 'ColorPanel';
ColorPicker.displayName = 'ColorPicker';

export default ColorPicker;
