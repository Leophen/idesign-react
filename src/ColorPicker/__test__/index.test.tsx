import { render, fireEvent } from '@testing-library/react';
import ColorPicker from '../index';

describe('ColorPicker 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<ColorPicker />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    const { container } = render(<ColorPicker value='#5cc083' />);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('background: #5cc083');
  });

  it('defaultValue', () => {
    const { container } = render(<ColorPicker value='#5cc083' />);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('background: #5cc083');
  });

  it('disabled', () => {
    const clickFn = jest.fn();
    const { container } = render(<ColorPicker disabled onTrigger={clickFn} />);
    expect(container.firstChild.firstChild.firstChild).toHaveClass('i-color__disabled');
    fireEvent.click(container.firstChild.firstChild.firstChild);
    expect(clickFn).toBeCalledTimes(0);
  });

  it('size', () => {
    const { container } = render(<ColorPicker size="small" />);
    expect(container.firstChild.firstChild.firstChild).toHaveClass('i-color--size-small');
  });
});

