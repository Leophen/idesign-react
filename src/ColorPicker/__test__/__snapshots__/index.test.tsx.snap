// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-color-picker"
  >
    <div
      class="i-popup__reference"
    >
      <div
        class="i-color"
        style="background: rgb(94, 98, 234);"
      />
    </div>
  </div>
</DocumentFragment>
`;
